File: src/cpp_basic_nodes/src/simple_service_server.cpp
/*
 * This program does the following
 *      1. Demonstrates accessing messages that have array types
 *      2. Demonstrates creating a service server
 */

// Include the ROS header file
#include <ros/ros.h>

// Include the header file for service
/*
 * These header files are generated after building messages, services and actions.
 * 
 * Local location: devel/include (inside the workspace folder)
 */
#include "cpp_basic_nodes/AddAllFloat64Numbers_cpp.h"

using namespace std;

// Declare a service server function
/*
 * Everytime the binded service is called, this function would be called. It takes two objects
 * a request and a response. The request will contain data, and response will have to be filled.
 * Since the objects are call by reference, the response is sent after a successful execution.
 * Return true if execution is successful.
 */
bool add_numbers(cpp_basic_nodes::AddAllFloat64Numbers_cpp::Request &req,
                 cpp_basic_nodes::AddAllFloat64Numbers_cpp::Response &res) {
    double ans = 0;  // Holds the sum
    for (const double &d: req.data) {   // Request is actually a vector type
        ans += d;
    }
    res.sum = ans;
    ROS_INFO_STREAM("Received " << req.data.size() << " numbers to add, result is " << ans);
    return true;
}

int main(int argc, char**argv) {
    // Initialize the node
    ros::init(argc, argv, "simple_cpp_service_server");
    // Create a NodeHandle
    ros::NodeHandle nh("~");
    
    // Create a service server object
    /*
     * This tells the master that this node advertises a service (it can serve it, and therefore be a "server").
     * The first argument is the name of the service. Since the nodehandle was created with the ~, the name will
     * be /simple_cpp_service_server/add_numbers
     * The second argument is the service server function. It is a callback funciton.
     * Note that just like subscriber callbacks, even service callbacks happen
     * 
     * Link: http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1NodeHandle.html#a20f789c8500181d91f8fecba4cd51cbc
     * Link (ServiceServer class): http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1ServiceServer.html
     */
    ros::ServiceServer srv_server = nh.advertiseService("add_numbers", add_numbers);
    
    // Info message
    /*
     * The ros::ServiceServer object can be used to access properties of the service
     * 
     * Link (ServiceServer class): http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1ServiceServer.html
     */
    ROS_INFO_STREAM("Service " << srv_server.getService() << " has been set up");
    // Set the entire thing ro run using spin
    ros::spin();
    
    return 0;
}

File: src/py_basic_nodes/scripts/simple_service_server.py
#!/usr/bin/env python3

"""
    Main objective of this program is to do the following:
        1. Demonstrates accessing messages that have array types
        2. Demonstrates creating a service server
"""

import rospy
import sys

# Import the module for using the service
"""
    This is developed after building the service files
    
    Local location: devel/lib/python3/dist-packages 
"""
from py_basic_nodes.srv import AddAllFloat64Numbers_py
from py_basic_nodes.srv import AddAllFloat64Numbers_pyRequest, AddAllFloat64Numbers_pyResponse


def add_numbers(req: AddAllFloat64Numbers_pyRequest):
    resp = AddAllFloat64Numbers_pyResponse()
    resp.sum = 0
    # Add all numbers
    for i in req.data:
        resp.sum += i
    rospy.loginfo("Received {0} numbers to add, result is {1}".format(len(req.data), resp.sum))
    return resp


def main():
    # Initialize node
    rospy.init_node("simple_py_service_server", argv=sys.argv, anonymous=False)

    # Create a service server
    """
        This tells the master that this node advertises a service (it can serve it, and therefore be a "server").
        The first argument is the service name
        The second argument is the service type (class)
        The third argument is the service server function. It is a callback function.
        
        Link: http://docs.ros.org/en/melodic/api/rospy/html/rospy.impl.tcpros_service.Service-class.html
    """
    srv_server = rospy.Service("{0}/add_numbers".format(rospy.get_name()), AddAllFloat64Numbers_py, add_numbers)

    # Info message
    rospy.loginfo("Service {0}/add_numbers has been started".format(rospy.get_name()))
    # Set the node to spin
    rospy.spin()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSException as e:
        rospy.logfatal("ROS error occurred: {0}".format(e))
